File: android/app/src/main/java/io/legado/app/ad/AdConfig.kt
package io.legado.app.ad

import com.google.gson.annotations.SerializedName

/**
 * 万象书屋广告配置.
 *
 * 与后端 `/api/ad-config` 返回的 `config` 字段结构一一对应, 也与
 * `assets/default_ad_config.json` 保持一致 (作为离线兜底).
 *
 * 设计要点:
 * - 三家 SDK 的真实 appId / posId **不在客户端硬编码**, 一律由后端下发,
 *   客户端默认空字符串. 这样万一某家账号被封, 后端改一行立刻切到备用账号或备用 SDK.
 * - [disabled] 是应急总开关, 任何场景下命中 true 都不出广告.
 * - [pollIntervalSec] 决定客户端多久拉一次远端配置.
 *
 * 兼容性: 字段都给了默认值, Gson 反序列化老/新版本不会崩.
 */
data class AdConfig(
    @SerializedName("disabled") val disabled: Boolean = true,
    @SerializedName("sdk") val sdk: SdkAppIds = SdkAppIds(),
    @SerializedName("placements") val placements: Placements = Placements(),
    @SerializedName("pollIntervalSec") val pollIntervalSec: Long = 6 * 3600L,
    // 万象书屋: 章节级付费墙. 头 N 章免费, 之后必须看广告解锁 M 分钟
    @SerializedName("chapterUnlock") val chapterUnlock: ChapterUnlock = ChapterUnlock()
) {
    data class SdkAppIds(
        @SerializedName("csj") val csj: SdkAppId = SdkAppId(),
        @SerializedName("ylh") val ylh: SdkAppId = SdkAppId()
    )

    data class SdkAppId(@SerializedName("appId") val appId: String = "")

    data class Placements(
        @SerializedName("splash") val splash: SplashPlacement = SplashPlacement(),
        @SerializedName("rewardedReadingUnlock")
        val rewardedReadingUnlock: RewardedPlacement = RewardedPlacement()
    )

    data class SplashPlacement(
        @SerializedName("enabled") val enabled: Boolean = false,
        @SerializedName("timeoutMs") val timeoutMs: Int = 3000,
        // 万象书屋: "独家投放" 应急开关. 后端在 getAdConfig 出口处把非选中 provider 的 weight 强制 0.
        // 空串 = 默认按 weight 抽签, "csj"/"ylh" = 仅这家. 优先级低于 breaker (熔断仍生效).
        @SerializedName("soloProvider") val soloProvider: String = "",
        @SerializedName("providers") val providers: List<ProviderSlot> = emptyList()
    )

    data class RewardedPlacement(
        @SerializedName("enabled") val enabled: Boolean = false,
        @SerializedName("unlockMinutes") val unlockMinutes: Int = 30,
        @SerializedName("cooldownMinutes") val cooldownMinutes: Int = 30,
        // 万象书屋累积奖励:
        // cooldownSec - 两次主动激励视频之间最小间隔, 防刷防 SDK 风控 (默认 180 = 3 分钟)
        // maxAccumulatedMinutes - 累积纯净阅读上限, 防恶意刷量 (默认 1440 = 24 小时)
        // showCountdownBar - 阅读器顶部是否显示倒计时条 + 看广告续期按钮
        @SerializedName("cooldownSec") val cooldownSec: Int = 180,
        @SerializedName("maxAccumulatedMinutes") val maxAccumulatedMinutes: Int = 1440,
        @SerializedName("showCountdownBar") val showCountdownBar: Boolean = true,
        // 万象书屋: 独家投放, 同 SplashPlacement.soloProvider
        @SerializedName("soloProvider") val soloProvider: String = "",
        @SerializedName("providers") val providers: List<ProviderSlot> = emptyList()
    )

    data class ProviderSlot(
        @SerializedName("name") val name: String,
        @SerializedName("weight") val weight: Int = 0,
        @SerializedName("posId") val posId: String = ""
    )

    /**
     * 万象书屋: 章节级付费墙. 用户读完头 [freeChapters] 章后必须看广告解锁;
     * 看完解锁 [unlockMinutes] 分钟内任意章节免费; 时间到再次拦截.
     *
     * 配合 RewardedPlacement 使用: 强制弹激励视频, 用户没看完 → 锁屏页 + 阻止翻下一章.
     */
    data class ChapterUnlock(
        @SerializedName("enabled") val enabled: Boolean = false,
        @SerializedName("freeChapters") val freeChapters: Int = 3,
        @SerializedName("unlockMinutes") val unlockMinutes: Int = 30,
        @SerializedName("blockOnSkip") val blockOnSkip: Boolean = true
    )

    /**
     * 是否对外完全不出广告: 总开关 / 拉到的远端表完全为空 都视为 disabled.
     *
     * 万象书屋: chapterUnlock 启用时即使 splash + rewarded 都关, 也不能算 effectivelyDisabled,
     * 否则付费墙的"立即弹激励"路径会因为 effectivelyDisabled=true 被短路, 导致用户被永久锁屏.
     * 反之, 任何"启用的功能"都视为 ad system 在工作.
     */
    fun effectivelyDisabled(): Boolean {
        if (disabled) return true
        val s = placements.splash
        val r = placements.rewardedReadingUnlock
        // 任意一个 placement 启用都不算 disabled
        return !s.enabled && !r.enabled && !chapterUnlock.enabled
    }
}

/** 后端 wrapping: { version, etag, config } */
data class AdConfigEnvelope(
    @SerializedName("version") val version: Long = 0,
    @SerializedName("etag") val etag: String = "",
    @SerializedName("config") val config: AdConfig = AdConfig()
)

/** 抽象出的"广告位类型", App 端只在这两个枚举值中触发 */
enum class AdPlacement(val key: String) {
    Splash("splash"),
    RewardedReadingUnlock("rewardedReadingUnlock"),
}

File: backend/models/adConfig.js
// 万象书屋: 广告配置 + 灰度 + 熔断检测

let db;

const DEFAULT_AD_CONFIG = {
  disabled: false,
  primary: 'csj',
  sdk: {
    csj: { appId: '', androidAppId: '5822340', iosAppId: '5825810' },
    ylh: { appId: '', androidAppId: '1217143733', iosAppId: '1217620900' }
  },
  placements: {
    splash: {
      enabled: true,
      timeoutMs: 3000,
      soloProvider: '',
      providers: [
        { name: 'csj', weight: 80, posId: '', androidPosId: '980622522', iosPosId: '981264244' },
        { name: 'ylh', weight: 20, posId: '', androidPosId: '7300219291644321', iosPosId: '4300170851839651' }
      ]
    },
    rewardedReadingUnlock: {
      enabled: true,
      unlockMinutes: 30,
      cooldownMinutes: 30,
      cooldownSec: 180,
      maxAccumulatedMinutes: 1440,
      showCountdownBar: true,
      soloProvider: '',
      providers: [
        { name: 'csj', weight: 80, posId: '', androidPosId: '980622521', iosPosId: '981263226' },
        { name: 'ylh', weight: 20, posId: '', androidPosId: '5330818271549483', iosPosId: '3310279811340675' }
      ]
    }
  },
  pollIntervalSec: 21600,
  chapterUnlock: {
    enabled: false,
    freeChapters: 3,
    unlockMinutes: 30,
    blockOnSkip: true
  }
};

function init(database) {
  db = database;
}

function _mergeAdConfigDefaults(saved, defaults) {
  if (saved == null || typeof saved !== 'object') return defaults;
  if (Array.isArray(saved)) return saved;
  const out = { ...saved };
  for (const k of Object.keys(defaults)) {
    if (out[k] === undefined) {
      out[k] = defaults[k];
    } else if (defaults[k] && typeof defaults[k] === 'object' && !Array.isArray(defaults[k])) {
      out[k] = _mergeAdConfigDefaults(out[k], defaults[k]);
    }
  }
  return out;
}

function _applySoloProvider(cfg) {
  if (!cfg || !cfg.placements) return;
  for (const placement of Object.values(cfg.placements)) {
    if (!placement || !Array.isArray(placement.providers)) continue;
    const solo = placement.soloProvider;
    if (!solo) continue;
    for (const slot of placement.providers) {
      if (slot.name !== solo) slot.weight = 0;
    }
  }
}

/**
 * 按平台解析 posId / appId:
 * - providers[].iosPosId / androidPosId → 覆盖 posId
 * - sdk.csj.iosAppId / androidAppId → 覆盖 appId
 * 客户端只读 posId / appId, 无需感知平台差异.
 */
function _resolvePlatformIds(cfg, platform) {
  if (!cfg) return;
  const key = platform === 'ios' ? 'iosPosId' : 'androidPosId';
  const appKey = platform === 'ios' ? 'iosAppId' : 'androidAppId';

  // sdk appId
  if (cfg.sdk) {
    for (const sdkCfg of Object.values(cfg.sdk)) {
      if (sdkCfg && sdkCfg[appKey]) {
        sdkCfg.appId = sdkCfg[appKey];
      }
    }
  }

  // placements posId
  if (cfg.placements) {
    for (const placement of Object.values(cfg.placements)) {
      if (!placement || !Array.isArray(placement.providers)) continue;
      for (const slot of placement.providers) {
        if (slot[key]) {
          slot.posId = slot[key];
        }
      }
    }
  }
}

function getAdConfig(deviceId, platform) {
  const row = db.prepare(
    'SELECT version, json, etag, staging_json, rollout_pct FROM ad_config WHERE id = 1'
  ).get();
  if (!row) {
    const cfg = JSON.parse(JSON.stringify(DEFAULT_AD_CONFIG));
    _resolvePlatformIds(cfg, platform || 'android');
    return { version: 0, json: JSON.stringify(cfg), etag: 'v0', isStaging: false };
  }

  let chosenJson = row.json;
  let chosenEtag = row.etag;
  let isStaging = false;

  const pct = row.rollout_pct || 0;
  if (pct > 0 && row.staging_json) {
    let inRollout = pct >= 100;
    if (!inRollout && deviceId) {
      const h = require('crypto').createHash('md5').update(deviceId).digest();
      const bucket = h.readUInt32LE(0) % 100;
      inRollout = bucket < pct;
    }
    if (inRollout) {
      chosenJson = row.staging_json;
      chosenEtag = row.etag + '-s' + pct;
      isStaging = true;
    }
  }

  try {
    const saved = JSON.parse(chosenJson);
    const merged = _mergeAdConfigDefaults(saved, DEFAULT_AD_CONFIG);
    _applySoloProvider(merged);
    _resolvePlatformIds(merged, platform || 'android');
    return { version: row.version, json: JSON.stringify(merged), etag: chosenEtag, isStaging };
  } catch (e) {
    console.error('[db.getAdConfig] JSON parse failed, falling back to DEFAULT_AD_CONFIG:', e.message);
    const cfg = JSON.parse(JSON.stringify(DEFAULT_AD_CONFIG));
    _resolvePlatformIds(cfg, platform || 'android');
    return {
      version: row.version,
      json: JSON.stringify(cfg),
      etag: 'fallback-' + (row.etag || ''),
      isStaging: false
    };
  }
}

function getAdConfigRaw() {
  const row = db.prepare('SELECT version, json, etag FROM ad_config WHERE id = 1').get();
  if (!row) {
    return { version: 0, json: JSON.stringify(DEFAULT_AD_CONFIG), etag: 'v0' };
  }
  try {
    const saved = JSON.parse(row.json);
    const merged = _mergeAdConfigDefaults(saved, DEFAULT_AD_CONFIG);
    return { version: row.version, json: JSON.stringify(merged), etag: row.etag };
  } catch (_) {
    return row;
  }
}

function saveAdConfig(jsonObj) {
  if (!jsonObj || typeof jsonObj !== 'object' || !jsonObj.placements) {
    throw new Error('ad config must contain placements');
  }
  const now = Date.now();
  const json = JSON.stringify(jsonObj);
  const cur = db.prepare('SELECT version FROM ad_config WHERE id = 1').get();
  const nextVer = (cur ? cur.version : 0) + 1;
  const etag = `v${nextVer}-${require('crypto').createHash('md5').update(json).digest('hex').slice(0, 8)}`;
  const tx = db.transaction(() => {
    db.prepare(
      `INSERT INTO ad_config(id, version, json, etag, updated_at) VALUES (1, ?, ?, ?, ?)
       ON CONFLICT(id) DO UPDATE SET version=excluded.version, json=excluded.json, etag=excluded.etag, updated_at=excluded.updated_at`
    ).run(nextVer, json, etag, now);
    db.prepare(
      'INSERT OR REPLACE INTO ad_config_history(version, json, created_at) VALUES (?, ?, ?)'
    ).run(nextVer, json, now);
    db.prepare(
      'DELETE FROM ad_config_history WHERE version <= (SELECT MAX(version) FROM ad_config_history) - 30'
    ).run();
  });
  tx();
  return { version: nextVer, etag };
}

function listAdConfigHistory(limit = 30) {
  return db.prepare('SELECT version, created_at FROM ad_config_history ORDER BY version DESC LIMIT ?').all(limit);
}

function getAdConfigByVersion(version) {
  return db.prepare('SELECT version, json, created_at FROM ad_config_history WHERE version = ?').get(version);
}

function setAdConfigStaging(jsonObj, rolloutPct) {
  if (!jsonObj || typeof jsonObj !== 'object' || !jsonObj.placements) {
    throw new Error('staging config must contain placements');
  }
  const pct = Math.max(0, Math.min(100, parseInt(rolloutPct, 10) || 0));
  db.prepare('UPDATE ad_config SET staging_json=?, rollout_pct=? WHERE id=1')
    .run(JSON.stringify(jsonObj), pct);
}

function commitAdConfigStaging() {
  const row = db.prepare('SELECT staging_json, rollout_pct FROM ad_config WHERE id = 1').get();
  if (!row || !row.staging_json) throw new Error('no staging config to commit');
  const obj = JSON.parse(row.staging_json);
  saveAdConfig(obj);
  db.prepare('UPDATE ad_config SET staging_json=NULL, rollout_pct=0 WHERE id=1').run();
}

function abortAdConfigStaging() {
  db.prepare('UPDATE ad_config SET staging_json=NULL, rollout_pct=0 WHERE id=1').run();
}

module.exports = {
  init, DEFAULT_AD_CONFIG,
  getAdConfig, getAdConfigRaw, saveAdConfig,
  listAdConfigHistory, getAdConfigByVersion,
  setAdConfigStaging, commitAdConfigStaging, abortAdConfigStaging,
};
